### ISSUE TITLE
S40RTS error in parallel

### ISSUE THREAD

**Issue by tjhei** [geodynamics/aspect#984]
If using MPI with >1 processor, error messages when data files are missing are not shown. Instead, the code hangs. Fix this.

**Comment by bangerth** [comment]
Got a little reproducer?

**Comment by bjs2** [comment]
After compiling the quick fix you made earlier today @tjhei, I ran the problem again with two processors. Although we already know the problem, I think the error message given when MPI aborts is still helpful:
```
bjs2@bjs2-Latitude-E6430:~/aspect/build$ mpirun -n 2 ./aspect ../benchmark/performance/rts_3d_example.prm 
-----------------------------------------------------------------------------
-- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
--     . version 1.5.0-pre
--     . running in DEBUG mode
--     . running with 2 MPI processes
--     . using Trilinos
-----------------------------------------------------------------------------

ARGH!
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI COMMUNICATOR 3 DUP FROM 0 
with errorcode 2.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
[1]PETSC ERROR: ------------------------------------------------------------------------
[1]PETSC ERROR: Caught signal number 15 Terminate: Some process (or the batch system) has told this process to end
[1]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[1]PETSC ERROR: or see http://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[1]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[1]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
[1]PETSC ERROR: to get more information on the crash.
[1]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[1]PETSC ERROR: Signal received
[1]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[1]PETSC ERROR: Petsc Release Version 3.6.4, Apr, 12, 2016 
[1]PETSC ERROR: ./aspect on a arch-linux2-c-opt named bjs2-Latitude-E6430 by bjs2 Wed Jun 29 14:49:58 2016
[1]PETSC ERROR: Configure options --prefix=/home/bjs2/libs/petsc-3.6.4 --with-debugging=0 --with-shared-libraries=1 --with-mpi=1 --with-x=0 --with-64-bit-indices=0 --download-hypre=1 CC=mpicc CXX=mpicxx FC=mpif90
[1]PETSC ERROR: #1 User provided function() line 0 in  unknown file
--------------------------------------------------------------------------
mpirun has exited due to process rank 0 with PID 4750 on
node bjs2-Latitude-E6430 exiting improperly. There are two reasons this could occur:

1. this process did not call "init" before exiting, but others in
the job did. This can cause a job to hang indefinitely while it waits
for all processes to call "init". By rule, if one process calls "init",
then ALL processes must call "init" prior to termination.

2. this process called "init", but exited without calling "finalize".
By rule, all processes that call "init" MUST call "finalize" prior to
exiting or it will be considered an "abnormal termination"

This may have caused other processes in the application to be
terminated by signals sent by mpirun (as reported here).
--------------------------------------------------------------------------
[bjs2-Latitude-E6430:04749] 1 more process has sent help message help-mpi-api.txt / mpi-abort
[bjs2-Latitude-E6430:04749] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```
It gives this message when I run it with just 1 processor as well, but that's because we now bypass the throw. But the message itself explains why the second processor refuses to terminate the job.

**Comment by bjs2** [comment]
The alteration that we made to utilities.cc works both when a file is not found and when the file size is 0. However, in the latter case, it gives the same error message as above.

**Comment by bangerth** [comment]
I'm still in favor of letting processor 0 read the file, communicate to everyone whether that succeeded, and only then trigger the AssertThrow. The other processors would still receive whether the reading worked, and if it didn't, throw a `SilentException`.

**Comment by bjs2** [comment]
I think that is what we're doing now, with some extra code to handle the case that the file exists but contains nothing. I haven't committed the changes yet because I wanted to be sure that this is the best way to go about things.
```
    std::string
    read_and_distribute_file_content(const std::string &filename,
                                     const MPI_Comm &comm)
    {
      std::string data_string;

      if (Utilities::MPI::this_mpi_process(comm) == 0)
        {
          // set file size to an invalid size (signalling an error if we can not read it)
          unsigned int filesize = numbers::invalid_unsigned_int;

          std::ifstream filestream(filename.c_str());
	        if (!filestream)
	          {
              // broadcast failure state, then throw
              MPI_Bcast(&filesize,1,MPI_UNSIGNED,0,comm);
              AssertThrow (false,
                           ExcMessage (std::string("Could not open file <") + filename + ">."));
              return data_string; // never reached
	          }

          // Read data from disk
          std::stringstream datastream;
          filestream >> datastream.rdbuf();

          if (!filestream.eof())
            {
              // broadcast failure state, then throw
              MPI_Bcast(&filesize,1,MPI_UNSIGNED,0,comm);
              AssertThrow (false,
                           ExcMessage (std::string("Reading of file ") + filename + " finished " +
                                   "before the end of file was reached. Is the file corrupted or"
                                   "too large for the input buffer?"));
              return data_string; // never reached
            }

          data_string = datastream.str();
          filesize = data_string.size();

          // Distribute data_size and data across processes
          MPI_Bcast(&filesize,1,MPI_UNSIGNED,0,comm);
          MPI_Bcast(&data_string[0],filesize,MPI_CHAR,0,comm);
        }
      else
        {
          // Prepare for receiving data
          unsigned int filesize;
          MPI_Bcast(&filesize,1,MPI_UNSIGNED,0,comm);
          if (filesize == numbers::invalid_unsigned_int)
            throw QuietException();
            
          data_string.resize(filesize);

          // Receive and store data
          MPI_Bcast(&data_string[0],filesize,MPI_CHAR,0,comm);
        }

      return data_string;
    }
```

**Comment by bangerth** [comment]
Yes, I think this is the correct way to go. Package it up!

**Comment by bjs2** [comment]
Resolved by #1075
